python/tests/shapes.py

```diff
@@ -0,0 +1,166 @@
+import unittest
+
+import ase
+import numpy as np
+
+import chemiscope
+
+from chemiscope.structures._shapes import SHAPE_PARAMS
+
+BASE_FRAME = ase.Atoms(numbers=np.zeros(3), positions=[[0, 0, 0], [1, 1, 1], [2, 2, 5]])
+BASE_FRAME.arrays["orientation"] = [[1, 0, 0, 0] for _ in BASE_FRAME]
+CUBE = [
+    [0, 0, 0],
+    [1, 0, 0],
+    [0, 1, 0],
+    [1, 1, 0],
+    [0, 0, 1],
+    [1, 0, 1],
+    [0, 1, 1],
+    [1, 1, 1],
+]
+CUBE_SPLX = [
+    [0, 1, 2],
+    [1, 2, 3],
+    [4, 5, 6],
+    [5, 6, 7],
+    [0, 1, 4],
+    [0, 2, 4],
+    [1, 3, 5],
+    [2, 3, 6],
+    [1, 4, 5],
+    [2, 4, 6],
+    [3, 5, 7],
+    [3, 6, 7],
+]
+
+SHAPE_DEFAULTS = {
+    "ellipsoid": {"semiaxes": [1, 2, 1]},
+    "polyhedron": {"vertices": CUBE, "simplices": CUBE_SPLX},
+    "sphere": {"radius": 1.0},
+}
+
+
+class TestStructures(unittest.TestCase):
+    """Conversion of structure data to chemiscope JSON"""
+
+    def test_no_shape(self):
+        frame = BASE_FRAME.copy()
+        data = chemiscope.create_input(frames=[frame])
+        self.assertTrue("shape" not in data["structures"][0])
+
+    def test_bad_shape(self):
+        frame = BASE_FRAME.copy()
+        frame.info["shape"] = "invalid"
+        with self.assertRaises(KeyError):
+            _ = chemiscope.create_input(frames=[frame])
+
+    def test_by_frame(self):
+        for key in SHAPE_DEFAULTS:
+            with self.subTest(shape=key):
+                frame = BASE_FRAME.copy()
+                frame.info["shape"] = key
+                for k, v in SHAPE_DEFAULTS[key].items():
+                    frame.info[k] = v
+                data = chemiscope.create_input(frames=[frame])
+                self.assertTrue("shape" in data["structures"][0])
+                self.assertTrue(
+                    all(
+                        [
+                            k in ss
+                            for s in data["structures"]
+                            for ss in s["shape"]["shape_1"]
+                            for k in SHAPE_PARAMS[key]["required"]
+                        ]
+                    )
+                )
+                self.assertTrue(
+                    all(
+                        [
+                            len(s["shape"]["shape_1"]) == len(frame)
+                            for s in data["structures"]
+                        ]
+                    )
+                )
+                for k in SHAPE_PARAMS[key]["optional"]:
+                    frame.info.pop(k, None)
+                    with self.assertWarns(Warning):
```

Comment:
same, could this check for the warning message?

Reply:
Done!